lib/Dialect/FIRRTL/FIRRTLOps.cpp

```diff
@@ -4724,6 +4728,72 @@ FIRRTLType RefSubOp::inferReturnType(ValueRange operands,
       loc, "ref.sub op requires a RefType of vector or bundle base type");
 }
 
+FIRRTLType RWProbeOp::inferReturnType(ValueRange operands,
+                                      ArrayRef<NamedAttribute> attrs,
+                                      std::optional<Location> loc) {
+  auto typeAttr = getAttr<TypeAttr>(attrs, "type");
+  auto type = typeAttr.getValue();
+  auto baseType = type_dyn_cast<FIRRTLBaseType>(type);
+  if (!baseType)
+    return emitInferRetTypeError(loc, "must be base type, not ", type);
+  if (baseType.hasUninferredWidth() || baseType.hasUninferredReset())
+    return emitInferRetTypeError(
+        loc, "must not have uninferred width or reset in type ", baseType);
+  auto forceableType = firrtl::detail::getForceableResultType(true, baseType);
+  if (!forceableType)
+    return emitInferRetTypeError(loc, "cannot force type ", type);
+  return forceableType;
+}
+
+LogicalResult RWProbeOp::verifyInnerRefs(hw::InnerRefNamespace &ns) {
+  auto targetRef = getTarget();
+  if (!targetRef)
+    return emitOpError("has invalid target reference");
+  if (targetRef.getModule() !=
+      (*this)->getParentOfType<FModuleLike>().getModuleNameAttr())
+    return emitOpError() << "has non-local target";
+
+  auto target = ns.lookup(targetRef);
+  if (!target)
+    return emitOpError() << "has target that cannot be resolved: " << target;
+
+  auto getFinalType = [&](auto type, auto fieldID,
+                          Location loc) -> FailureOr<Type> {
+    auto fieldIDType = type_dyn_cast<hw::FieldIDTypeInterface>(type);
+    if (fieldIDType)
+      return fieldIDType.getFinalTypeByFieldID(fieldID);
+    // Verifier on target op should handle this, check to be safe anyway.
+    if (fieldID != 0) {
+      auto diag =
+          emitOpError()
+          << "has target that resolves to invalid per-field inner symbol";
+      return diag.attachNote(loc)
+             << "target does not support per-field inner symbols";
+    }
```

Comment:
This would be better as an `assert`. Duplicating the checking (in production code) seems unnecessary.